src/expr/src/scalar/func/macros.rs

```diff
@@ -0,0 +1,127 @@
+// Copyright Materialize, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+macro_rules! fn_unary {
+    ($fn_name:ident($param:ident: $ty:ty) -> $ret:ty $body:block) => {
+        fn_unary!(stringify!($fn_name), $fn_name($param: $ty) -> $ret $body);
+    };
+
+    ($name:expr, $fn_name:ident($param:ident: $ty:ty) -> $ret:ty $body:block) => {
```

Comment:
How can we express the return type of for functions that return `ScalarType` values that don't directly correspond to primitive values in Rust? e.g. `ScalarType::Char`

Reply:
Same as above, it would be a manual implementation of the trait. However, I think for some types we can just define a corresponding Rust type. Like, for `ScalarType::Char` we can have a `struct Char(String)` as the Rust equivalent so you can write your function implementation based on that